Dependencies/xr/Source/ARKit/XR.mm

```diff
@@ -0,0 +1,782 @@
+*/
+- (void)updateFrameBuffer:(CVPixelBufferRef)arCameraBuffer
+{
+    // Lock the frame buffer with read access to stop ARKit from updating the camera buffer during update.
+    CVReturn ret = CVPixelBufferLockBaseAddress(arCameraBuffer, kCVPixelBufferLock_ReadOnly);
+    if (ret != kCVReturnSuccess)
+    {
+        return;
+    }
+    
+    @try
+    {
+        // Find the width, height, and format of the AR image.
+        size_t bufferWidth = CVPixelBufferGetWidth(arCameraBuffer);
+        size_t bufferHeight = CVPixelBufferGetHeight(arCameraBuffer);
+        auto format = CVPixelBufferGetPixelFormatType(arCameraBuffer);
+        
+        // If we don't have two planes the arCamera has not yet been initialized so back out for now.
+        if (CVPixelBufferGetPlaneCount(arCameraBuffer) < 2)
+        {
+            CVPixelBufferUnlockBaseAddress(arCameraBuffer, kCVPixelBufferLock_ReadOnly);
+            return;
+        }
+
+        // Check if the size of the frame buffer has changed, if so dispose of the old one and create a new buffer.
+        if (frameBuffer == nil || bufferWidth != CVPixelBufferGetWidth(frameBuffer) || bufferHeight !=CVPixelBufferGetHeight(frameBuffer))
+        {
+            if (frameBuffer != nil)
+            {
+                CVPixelBufferRelease(frameBuffer);
+                frameBuffer = nil;
+            }
+            
+            // We must specify PixelBufferMetalCompatibility on the frame buffer to make this composable into a Metal Texture.
+            auto attributes = [NSDictionary dictionaryWithObjectsAndKeys:[NSNumber numberWithBool:YES], kCVPixelBufferMetalCompatibilityKey, nil];
+            
+            // Create the frame buffer.
+            ret = CVPixelBufferCreate(kCFAllocatorDefault, bufferWidth, bufferHeight, format, (__bridge CFDictionaryRef)attributes, &frameBuffer);
+            if (ret != kCVReturnSuccess)
+            {
+                throw std::runtime_error("Failed to allocate frame buffer.");
+            }
+        }
+        
+        // Lock the pixel buffer for write access.
+        ret = CVPixelBufferLockBaseAddress(frameBuffer, 0);
+        if (ret != kCVReturnSuccess)
+        {
+            return;
+        }
+        
+        @try
+        {
+            // Copy both planes of the AR camera image to the frame buffer.
+            void* ydestPlane = CVPixelBufferGetBaseAddressOfPlane(frameBuffer, 0);
+            void* ysrcPlane = CVPixelBufferGetBaseAddressOfPlane(arCameraBuffer, 0);
+            memcpy(ydestPlane, ysrcPlane, bufferWidth * bufferHeight);
+            
+            auto uvdestPlane = CVPixelBufferGetBaseAddressOfPlane(frameBuffer, 1);
+            auto uvsrcPlane = CVPixelBufferGetBaseAddressOfPlane(arCameraBuffer, 1);
+            memcpy(uvdestPlane, uvsrcPlane, bufferWidth * bufferHeight / 2);
+        }
+        @finally
+        {
+            CVPixelBufferUnlockBaseAddress(arCameraBuffer, 0);
```

Comment:
Are these two finally blocks reversed? I'd have though the outer one would free the `arCameraBuffer` lock.

Reply:
Yes fixed